pkg/sql/mvcc_statistics_update_job.go

```diff
@@ -17,40 +17,141 @@ package sql
 
-func (j *mvccStatisticsUpdateJob) Resume(ctx context.Context, execCtxI interface{}) (jobErr error) {
+// Resume implements the jobs.Resumer interface.
+func (j *mvccStatisticsUpdateJob) Resume(ctx context.Context, execCtxI interface{}) error {
+	log.Infof(ctx, "starting mvcc statistics update job")
+
+	// This job is a forever running background job, and it is always safe to
+	// terminate the SQL pod whenever the job is running, so mark it as idle.
+	j.job.MarkIdle(true)
+
+	execCtx := execCtxI.(JobExecContext)
+
+	// Export global metrics for tenants if this is an out-of-process SQL node.
+	// All external mode tenant servers have no node IDs.
+	if _, hasNodeID := execCtx.ExecCfg().NodeInfo.NodeID.OptionalNodeID(); !hasNodeID {
+		return j.runTenantGlobalMetricsExporter(ctx, execCtx)
+	}
+
 	// TODO(zachlite):
 	// Delete samples older than configurable setting...
 	// Collect span stats for tenant descriptors...
 	// Write new samples...
-	execCtx := execCtxI.(JobExecContext)
-	stopper := execCtx.ExecCfg().DistSQLSrv.Stopper
-	j.job.MarkIdle(true)
+
+	// Block until context is cancelled since there's nothing that needs to be
+	// done here. We should not return nil, or else the job will be marked as
+	// succeeded.
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+// runTenantGlobalMetricsExporter executes the logic to export global metrics
+// for tenants.
+func (j *mvccStatisticsUpdateJob) runTenantGlobalMetricsExporter(
+	ctx context.Context, execCtx JobExecContext,
+) error {
+	metricsRegistry := execCtx.ExecCfg().MetricsRecorder.AppRegistry()
+
+	initialRun := true
+	defer func() {
+		metricsRegistry.RemoveMetric(j.dynamicMetrics.livebytes)
+	}()
+
+	runTask := func() error {
+		resp, err := execCtx.ExecCfg().TenantStatusServer.SpanStats(
+			ctx,
+			&roachpb.SpanStatsRequest{
+				// Fan out to all nodes. SpanStats takes care of only contacting
+				// the relevant nodes with the tenant's span.
+				NodeID: "0",
+				Spans:  []roachpb.Span{execCtx.ExecCfg().Codec.TenantSpan()},
+			},
+		)
+		if err != nil {
+			return err
+		}
+		var total int64
+		for _, stats := range resp.SpanToStats {
+			total += stats.ApproximateTotalStats.LiveBytes
+		}
+		j.dynamicMetrics.livebytes.Update(total)
+
+		// Only register metrics once we get our initial values. This avoids
+		// metrics from fluctuating whenever the job restarts.
+		if initialRun {
+			metricsRegistry.AddMetric(j.dynamicMetrics.livebytes)
+			initialRun = false
+		}
+		return nil
+	}
+
+	timer := timeutil.NewTimer()
+	defer timer.Stop()
+
+	// Fire the timer immediately to start the initial update.
+	timer.Reset(0)
 
 	for {
 		select {
 		case <-ctx.Done():
-			return nil
-
-		case <-stopper.ShouldQuiesce():
-			return nil
+			return ctx.Err()
+		case <-timer.C:
+			timer.Read = true
+			timer.Reset(TenantGlobalMetricsExporterInterval.Get(&execCtx.ExecCfg().Settings.SV))
+			if err := runTask(); err != nil {
+				log.Errorf(ctx, "mvcc statistics update job error: %v", err)
```

Comment:
Do we definitely want to Reset the timer before running the task? If the task takes > a minute in some clusters it'll mean that we are more or less always running the task.  Perhaps not a real issue in practice.

Reply:
This shouldn't be a real issue in practice, especially for Serverless clusters. Even if that becomes a problem, we could increase the interval via a host override. Regardless, I went ahead and made that change. The metric will now only be updated at least 1 minute after the last update.